mypy/build.py

```diff
@@ -891,36 +932,57 @@ def find() -> Optional[str]:
                 if isdir:
                     dirs.append(dir)
             find_module_dir_cache[dir_chain, lib_path] = dirs
-        candidate_base_dirs = find_module_dir_cache[dir_chain, lib_path]
-
+        third_party_dirs = []
+        # Third-party stub/typed packages
+        for pkg_dir in site_packages_dirs:
+            stub_name = components[0] + '-stubs'
+            typed_file = os.path.join(pkg_dir, components[0], 'py.typed')
+            stub_dir = os.path.join(pkg_dir, stub_name)
+            if os.path.isdir(stub_dir):
+                stub_components = [stub_name] + components[1:]
+                path = os.path.join(pkg_dir, *stub_components[:-1])
+                if os.path.isdir(path):
+                    third_party_dirs.append(path)
+            elif os.path.isfile(typed_file):
+                path = os.path.join(pkg_dir, dir_chain)
+                third_party_dirs.append(path)
+
+        candidate_base_dirs = third_party_dirs + find_module_dir_cache[dir_chain, lib_path]
         # If we're looking for a module like 'foo.bar.baz', then candidate_base_dirs now
         # contains just the subdirectories 'foo/bar' that actually exist under the
         # elements of lib_path.  This is probably much shorter than lib_path itself.
         # Now just look for 'baz.pyi', 'baz/__init__.py', etc., inside those directories.
-        seplast = os.sep + components[-1]  # so e.g. '/baz'
+        seplast = os.sep + components[-1]
         sepinit = os.sep + '__init__'
         for base_dir in candidate_base_dirs:
             base_path = base_dir + seplast  # so e.g. '/usr/lib/python3.4/foo/bar/baz'
             # Prefer package over module, i.e. baz/__init__.py* over baz.py*.
             for extension in PYTHON_EXTENSIONS:
                 path = base_path + sepinit + extension
+                path_stubs = base_path + '-stubs' + sepinit + extension
                 if is_file(path) and verify_module(id, path):
                     return path
+                elif is_file(path_stubs) and verify_module(id, path_stubs):
+                    return path_stubs
             # No package, look for module.
             for extension in PYTHON_EXTENSIONS:
                 path = base_path + extension
                 if is_file(path) and verify_module(id, path):
                     return path
         return None
 
-    key = (id, lib_path)
+    key = id, python_executable, lib_path
     if key not in find_module_cache:
+        components = id.split('.')
+        dir_chain = os.sep.join(components[:-1])  # e.g., 'foo/bar'
+        site_packages_dirs = get_site_packages_dirs(python_executable)
```

Comment:
Is there any reason not to put these three lines inside `find`, which makes the flow a little clearer?

Reply:
Not really. They originally were broken out due to the refactor, but I agree putting them in find is ideal right now.